server/common/immutable_kvcache.py

```diff
@@ -11,14 +11,44 @@ class ImmutableKVCache(MutableMapping):
     def __init__(self, factory):
         self.factory = factory
         self.lock = threading.Lock()
+        self.factory_calls = {}
         self.cache = {}
         super().__init__()
 
     def __getitem__(self, key):
-        if key not in self.cache:
-            with self.lock:
-                if key not in self.cache:
+        if key in self.cache:
+            return self.cache[key]
+
+        # we need to call factory.  First grab the main lock and the per-key CV.
+        factory_calls = None
+        with self.lock:
+            if key in self.cache:
+                return self.cache[key]
+            if key not in self.factory_calls:
+                creation_thr = True
+                self.factory_calls[key] = {'cv': threading.Condition(), 'is_done': False, 'error': None}
+            factory_calls = self.factory_calls[key]
+
+        # with the CV, create the value (or wait for it to be created)
+        cv = factory_calls['cv']
+        with cv:
+            if creation_thr:
```

Comment:
Is create_thr guaranteed to be set.  For example if key was in self.factor_calls (line 27) it might not be.   You may need a create_thr=False somewhere.

Reply:
fixed!  good eye.